storage/src/index/storage.rs

```diff
@@ -116,16 +248,18 @@ impl<T: Translator, V> Index<T, V> {
         self.map.is_empty()
     }
 
-    /// Insert a new record into the index.
-    pub fn insert(&mut self, key: &[u8], value: V) {
+    /// Insert a new record into the index. If there is a collision, then the new value will be
+    /// added in front of the rest of values for this translated key.
+    pub fn insert(&mut self, key: &[u8], mut value: V) {
         let translated_key = self.translator.transform(key);
 
         match self.map.entry(translated_key) {
-            Entry::Occupied(entry) => {
-                let entry: &mut Record<V> = entry.into_mut();
-                entry.next = Some(Box::new(Record {
+            Entry::Occupied(mut occupied_entry) => {
+                let record = occupied_entry.get_mut();
+                swap(&mut record.value, &mut value); // puts the new value at the front
```

Comment:
I'm not sure what this gives us? Seems like an extra op where we could otherwise just use `value` in the new Record without issue?

Reply:
It keeps the insertion order (wether we insert here or the iterator) LIFO, making the behavior more easily understandable. I figured it was worth this small bit of overhead. Or are you suggesting there may be an more efficient way to achieve this behavior?

Reply:
Described now in comments.